### PR TITLE
feat: support shell notification

### PR DESCRIPTION
**Pull request by haoel**
EaseProbe set the following environment variables
    - `$EASEPROBE_TYPE`:  two types of message:`Status` or `SLA`

If the TYPE is Status, the following environment variables would be set
    - `$EASEPROBE_NAME`: probe name
    - `$EASEPROBE_STATUS`: "up" or "down"
    - `$EASEPROBE_RTT`: round trip time in milliseconds
    - `$EASEPROBE_TIMESTAMP`: timestamp of probe time
    - `$EASEPROBE_MESSAGE`: probe the message

And EaseProbe offers two formats for both probe status and SLA report
    - `$EASEPROBE_JSON`: the JSON format
    - `$EASEPROBE_CSV`: the CSV format

The CVS format would be set to STDIN for the shell command. 

### Probe Status

```
Type: Status
----------------------------------------------------
Title: TLS Test Recovery - ( 19m24s Downtime )
Name: TLS Test
Status: up
RTT: 42000000
TIMESTAMP: 1654588743532
Message: - Success (http): HTTP Status Code is 200
----------------------------------------------------
{"name":"TLS Test Recovery - ( 19m24s Downtime )","endpoint":"https://localhost:8443/hello","time":"2022-06-07T07:59:03.532214Z","timestamp":1654588743532,"rtt":42418000,"status":"up","prestatus":"down","message":"Success (http): HTTP Status Code is 200"}
----------------------------------------------------
Title, Name, Endpoint, Status, PreStatus, RoundTripTime, Time, Timestamp, Message
TLS Test Recovery - ( 1m48s Downtime ), TLS Test, https://localhost:8443/hello, up, down, 43000000, 2022-06-08 05:26:16 UTC, 1654665976106, Success (http): HTTP Status Code is 200
---------------------------------------------------- 
```

### SLA Report
the output of `resources/scripts/notify/notify.sh`

```
Type: SLA
----------------------------------------------------
[{"name":"dummy","endpoint":"http://localhost:12345/dummay","sla":{"up":0,"down":615000000000,"sla":0},"probe_summary":{"total":41,"up":0,"down":41},"latest_probe":{"time":"2022-06-07T07:58:48.530724Z","status":"down","message":"Error (http): Error: Get \"http://localhost:12345/dummay\": dial tcp [::1]:12345: connect: connection refused"}},{"name":"MegaCloud","endpoint":"https://example.com/","sla":{"up":600000000000,"down":0,"sla":100},"probe_summary":{"total":20,"up":20,"down":0},"latest_probe":{"time":"2022-06-07T07:58:48.530868Z","status":"up","message":"Success (http): HTTP Status Code is 200"}},{"name":"MegaEase Website (China)","endpoint":"https://megaease.cn","sla":{"up":450000000000,"down":0,"sla":100},"probe_summary":{"total":30,"up":30,"down":0},"latest_probe":{"time":"2022-06-07T07:58:48.530705Z","status":"up","message":"Success (http): HTTP Status Code is 200"}},{"name":"TLS Test","endpoint":"https://localhost:8443/hello","sla":{"up":135000000000,"down":435000000000,"sla":23.684210526315788},"probe_summary":{"total":38,"up":9,"down":29},"latest_probe":{"time":"2022-06-07T07:58:48.530763Z","status":"down","message":"Error (http): Error: Get \"https://localhost:8443/hello\": dial tcp [::1]:8443: connect: connection refused"}}]
----------------------------------------------------
Name, Endpoint, UpTime, DownTime, SLA, ProbeSummary, LatestProbe, LatestStatus, Message
dummy, http://localhost:12345/dummay, up(0s), down(10m15s), 0.00%, 41(down : 41), 2022-06-07 07:58:48 UTC, down, Error (http): Error: Get "http://localhost:12345/dummay": dial tcp [::1]:12345: connect: connection refused
MegaCloud, https://example.com/, up(10m0s), down(0s), 100.00%, 20(up : 20), 2022-06-07 07:58:48 UTC, up, Success (http): HTTP Status Code is 200
MegaEase Website (China), https://megaease.cn, up(7m30s), down(0s), 100.00%, 30(up : 30), 2022-06-07 07:58:48 UTC, up, Success (http): HTTP Status Code is 200
TLS Test, https://localhost:8443/hello, up(2m15s), down(7m15s), 23.68%, 38(down : 29    up : 9), 2022-06-07 07:58:48 UTC, down, Error (http): Error: Get "https://localhost:8443/hello": dial tcp [::1]:8443: connect: connection refused
---------------------------------------------------- 
```

### CODE REVIEW

**Comment by nullsimon** [comment]
For the CSV format output, I don't know if it needs CSV table header or not. Actually I will use json format instead, hahaha~

```
name,endpoint,sla/up,sla/down,sla/sla,probe_summary/total,probe_summary/up,probe_summary/down,latest_probe/time,latest_probe/status,latest_probe/message
dummy,http://localhost:12345/dummay,0,615000000000,0,41,0,41,2022-06-07T07:58:48.530724Z,down,"Error (http): Error: Get ""http://localhost:12345/dummay"": dial tcp [::1]:12345: connect: connection refused"
MegaCloud,https://example.com/,600000000000,0,100,20,20,0,2022-06-07T07:58:48.530868Z,up,Success (http): HTTP Status Code is 200
MegaEase Website (China),https://megaease.cn,450000000000,0,100,30,30,0,2022-06-07T07:58:48.530705Z,up,Success (http): HTTP Status Code is 200
TLS Test,https://localhost:8443/hello,135000000000,435000000000,23.684210526315788,38,9,29,2022-06-07T07:58:48.530763Z,down,"Error (http): Error: Get ""https://localhost:8443/hello"": dial tcp [::1]:8443: connect: connection refused"

```

File: notify/shell/shell.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+	}
+	cmd.Stdin = strings.NewReader(envMap["EASEPROBE_CSV"] + "\n")
+	cmd.Env = append(os.Environ(), env...)
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return err
+	}
+	log.Debugf("[%s / %s] - %s", c.NotifyKind, c.NotifyName, global.CommandLine(c.Cmd, c.Args))
+	log.Debugf("input: \n%s", msg)
+	log.Debugf("output:\n%s", string(output))
```
Comment (jordy1024) [inline notify/shell/shell.go:77]: redundant type conversion to string ? prefer?
```
log.Debugf("output:\n%s",output)
```

File: notify/shell/shell.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+type NotifyConfig struct {
+	base.DefaultNotify `yaml:",inline"`
+
+	Cmd  string   `yaml:"cmd"`
+	Args []string `yaml:"args"`
+}
+
+// Config is the config for shell probe
+func (c *NotifyConfig) Config(gConf global.NotifySettings) error {
+	c.NotifyKind = "shell"
```
Comment (jordy1024) [inline notify/shell/shell.go:45]: "shell" appears multiple times,suggest to to declare as a constant string

File: notify/shell/shell.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+	c.NotifyFormat = report.Shell
+	c.NotifySendFunc = c.RunShell
+	c.DefaultNotify.Config(gConf)
+
+	return nil
+}
+
+// RunShell is the shell for shell notify
+func (c *NotifyConfig) RunShell(title, msg string) error {
```
Comment (jordy1024) [inline notify/shell/shell.go:54]: redundant  parameter:
`title` declared but not used?

File: report/sla.go [hunk 4]
```diff
@@ -404,3 +404,38 @@ func SLASummary(probers []probe.Prober) string {
 	summary += "\n" + global.FooterString()
 	return summary
 }
+
+// SLACSVSection set the CSV format for SLA
+func SLACSVSection(r *probe.Result) string {
+	text := "%s, %s, up(%s), down(%s), %.2f%%, %d(%s), %s, %s, %s"
+	return fmt.Sprintf(text, r.Name, r.Endpoint,
```
Comment (jordy1024) [inline report/sla.go:411]: prefer ? 
```
return fmt.Sprint(
```

over 

```
return fmt.Sprintf(
```

File: notify/shell/shell.go (hunk 1 above)
Comment (haoel) [reply notify/shell/shell.go:77]: "%s" would cast the []byte to string implicitly,  using explicit cast would let the people understand the output is not the string. Personally, I think the explicit cast can improve the code quality.

File: notify/shell/shell.go (hunk 3 above)
Comment (haoel) [reply notify/shell/shell.go:54]: this function is a callback function, for some notifications like email, it needs the title, but for others like sms, the title is useless.

File: notify/shell/shell.go (hunk 2 above)
Comment (haoel) [reply notify/shell/shell.go:45]: actually, it only appears once.

File: report/sla.go (hunk 4 above)
Comment (haoel) [reply report/sla.go:411]: here, using Sprintf instead of Sprint is better. and using Sprintf way can be easy to maintain.

**Comment by haoel** [comment]
> For the CSV format output, I don't know if it needs CSV table header or not. Actually I will use json format instead, hahaha~
it makes sense to supply a head, I will do that.

File: global/global.go [hunk 5]
```diff
@@ -214,3 +214,12 @@ func MakeDirectory(filename string) string {
 
 	return filepath.Join(dir, file)
 }
+
+// CommandLine will return the whole command line which includes command and all arguments
+func CommandLine(cmd string, args []string) string {
+	result := cmd
+	for _, arg := range args {
+		result += " " + arg
```
Comment (localvar) [inline global/global.go:222]: the result string could be wrong if `arg` contains one or more spaces

File: report/sla.go [hunk 6]
```diff
@@ -404,3 +404,38 @@ func SLASummary(probers []probe.Prober) string {
...
+		r.Stat.Total, SLAStatusText(r.Stat, Text),
+		r.StartTime.UTC().Format(r.TimeFormat),
+		r.Status.String(), r.Message)
+}
+
+// SLACSV return a full stat report with CSV format
+func SLACSV(probers []probe.Prober) string {
+	csv := "Name, Endpoint, UpTime, DownTime, SLA, ProbeSummary, LatestProbe, LatestStatus, Message\n"
+	for _, p := range probers {
+		csv += SLACSVSection(p.Result()) + "\n"
```
Comment (localvar) [inline report/sla.go:422]: propose to use `bytes.Buffer` to improve performance in case there are many probers.

Comment (localvar) [reply report/sla.go:422]: and propose to use `encoding/csv` to generate the CSV result, it handles corner cases better.

File: README.md [hunk 7]
```diff
@@ -777,6 +785,24 @@ notify:
...
+  # EaseProbe set the following environment variables
+  #  - TYPE: "Status" or "SLA"
+  #  - NAME: probe name
+  #  - STATUS: "up" or "down"
+  #  - RTT: round trip time in milliseconds
+  #  - TIMESTAMP: timestamp of probe time
+  #  - MESSAGE: probe message
+  # and offer two formats of string
+  #  - JSON: the JSON format
+  #  - CSV: the CSV format
```
Comment (proditis) [inline README.md:797]:
```suggestion
  #  - EASEPROBE_TYPE: "Status" or "SLA"
  #  - EASEPROBE_NAME: probe name
  #  - EASEPROBE_STATUS: "up" or "down"
  #  - EASEPROBE_RTT: round trip time in milliseconds
  #  - EASEPROBE_TIMESTAMP: timestamp of probe time
  #  - EASEPROBE_MESSAGE: probe message
  # and offer two formats of string
  #  - EASEPROBE_JSON: the JSON format
  #  - EASEPROBE_CSV: the CSV format
```

File: probe/shell/shell.go [hunk 8]
```diff
@@ -66,12 +65,8 @@ func (s *Shell) DoProbe() (bool, string) {
 	ctx, cancel := context.WithTimeout(context.Background(), s.ProbeTimeout)
 	defer cancel()
 
-	for _, e := range s.Env {
-		v := strings.Split(e, "=")
-		os.Setenv(v[0], v[1])
-	}
-
 	cmd := exec.CommandContext(ctx, s.Command, s.Args...)
+	cmd.Env = append(os.Environ(), s.Env...)
```
Comment (proditis) [inline probe/shell/shell.go:69]: This gets the environment of the easeprobe executable and appends our new Env (status,name,rtt etc)? Is so this is not advisable. Just set the new environment variables and let the system inherit/propagate the environment variables based on their export criteria...

Comment (haoel) [reply probe/shell/shell.go:69]: I did some tests and read the source code.
for Go StdLib,
- if we define the `cmd.Env`, it only sets those env variables, the parent's or system env won't be set.
- if the `cmd.Env` is not defined. then use the system default env, which is the `os.Environ()`

So, if we need to pass the system env to the child process, we have to do `cmd.Env := append(os.Environ(), env…)`

I think the current implementation is right.

Comment (jordy1024) [reply probe/shell/shell.go:69]: It is a clear and thorough analysis.👍

Comment (jordy1024) [reply probe/shell/shell.go:69]: > I did some tests and read the source code. for Go StdLib,
> 
> * if we define the `cmd.Env`, it only sets those env variables, the parent's or system env won't be set.
> * if the `cmd.Env` is not defined. then use the system default env, which is the `os.Environ()`
> 
> So, if we need to pass the system env to the child process, we have to do `cmd.Env := append(os.Environ(), env…)`
> 
> I think the current implementation is right.

It is a clear and thorough analysis.👍 @haoel

**Pull request merged**